File: src/configuration.ts
/** @fileoverview Interprets project configuration and source-scoped exception policies in deterministic validation order. */

import type {
  LibraryRef,
  ProjectConfig,
  RuleReplacement,
  LibrarySource,
} from './configuration-types';
import {
  compare,
  field,
  groupId,
  invalid,
  nonempty,
  object,
  relativePath,
  ruleGroup,
  strings,
} from './formats/validation';

/** Reject unknown own fields before interpreting a configuration object. */
function knownFields(
  value: Record<string, unknown>,
  allowed: ReadonlyArray<string>,
  location: string,
): void {
  for (const name of Object.keys(value)) {
    if (!allowed.includes(name)) invalid(location, `unknown field ${name}`);
  }
}

/** Validate and register a repository before checking the source's later fields; duplicate names are case-insensitive. */
function sourceRepository(
  value: unknown,
  where: string,
  repositories: Set<string>,
): string {
  const repository = nonempty(value, `${where}.repository`);
  if (
    !/^[A-Za-z0-9][A-Za-z0-9-]*\/[A-Za-z0-9_.-]+$/u.test(repository) ||
    /\/(\.|\.\.)$/u.test(repository)
  )
    return invalid(where, 'repository must use owner/name form');
  if (repositories.has(repository.toLowerCase()))
    return invalid(
      where,
      `repository ${repository} is declared more than once`,
    );
  repositories.add(repository.toLowerCase());
  return repository;
}

/** Classify a full commit or exact tag, rejecting invalid Git ref components without I/O. */
function sourceRef(ref: string, where: string): LibraryRef {
  if (/^[a-f0-9]{40}$/iu.test(ref))
    return { kind: 'commit', sha: ref.toLowerCase() };
  const tag = ref.startsWith('refs/tags/') ? ref.slice(10) : ref;
  if (
    /[\x00-\x20\x7f~^:?*\[\\]|\.\.|@\{/u.test(tag) ||
    tag === '@' ||
    tag.startsWith('-') ||
    (ref.startsWith('refs/') && !ref.startsWith('refs/tags/')) ||
    tag
      .split('/')
      .some(
        (part) =>
          !part ||
          part.startsWith('.') ||
          part.endsWith('.lock') ||
          part.endsWith('.'),
      )
  )
    return invalid(
      `${where}.ref`,
      'expected a full commit SHA or exact tag name',
    );
  if (/^[a-f0-9]{4,39}$/iu.test(ref))
    return invalid(
      `${where}.ref`,
      'abbreviated commits are unsupported; use a full SHA or refs/tags/<name>',
    );
  return { kind: 'tag', name: `refs/tags/${tag}` };
}

/** Read nonempty exclusion reasons in sorted rule-ID order. */
function exclusionDecisions(
  value: unknown,
  where: string,
): ReadonlyMap<string, string> {
  const exclude = new Map(
    Object.entries(object(value, `${where}.exclude`))
      .sort(([a], [b]) => compare(a, b))
      .map(([id, reason]) => [id, nonempty(reason, `${where}.exclude.${id}`)]),
  );
  return exclude;
}

/** Validate one replacement's local file and reason without resolving the referenced file. */
function replacementDeclaration(
  rawReplacement: unknown,
  id: string,
  where: string,
): RuleReplacement {
  const replacement = object(rawReplacement, `${where}.replace.${id}`);
  knownFields(replacement, ['file', 'reason'], `${where}.replace.${id}`);
  const location = `${where}.replace.${id}.file`;
  const text = nonempty(field(replacement, 'file'), location);
  const path = relativePath(text, location);
  if (!path.startsWith('local/'))
    return invalid(location, 'replacement files must be under local/');
  return {
    file: path,
    reason: nonempty(
      field(replacement, 'reason'),
      `${where}.replace.${id}.reason`,
    ),
  };
}

/** Read replacement declarations in sorted target-ID order. */
function replacementDecisions(
  value: unknown,
  where: string,
): ReadonlyMap<string, RuleReplacement> {
  const replace = new Map<string, RuleReplacement>(
    Object.entries(object(value, `${where}.replace`))
      .sort(([a], [b]) => compare(a, b))
      .map(([id, rawReplacement]) => [
        id,
        replacementDeclaration(rawReplacement, id, where),
      ]),
  );
  return replace;
}

/** Validate distinct group IDs at their original array indices before sorting for resolution. */
function selectedGroups(
  value: unknown,
  location: string,
): ReadonlyArray<string> {
  return strings(value, location)
    .map((id, index) => groupId(id, `${location}[${index}]`))
    .sort(compare);
}

/** Parse a source's fields in declaration-policy order, then reject contradictory exceptions. */
function sourceConfiguration(
  name: string,
  raw: unknown,
  repositories: Set<string>,
): LibrarySource {
  const where = `sources.${name}`;
  if (!/^[a-z][a-z0-9-]*$/u.test(name) || name === 'local')
    return invalid(where, 'invalid or reserved source name');
  const source = object(raw, where);
  knownFields(
    source,
    ['repository', 'ref', 'groups', 'exclude', 'replace'],
    where,
  );
  const repository = sourceRepository(
    field(source, 'repository'),
    where,
    repositories,
  );
  const ref = nonempty(field(source, 'ref'), `${where}.ref`);
  const parsedRef = sourceRef(ref, where);
  const requestedGroups = field(source, 'groups');
  const groups =
    requestedGroups === '*' ||
    requestedGroups === 'techs/*' ||
    requestedGroups === 'practices/*'
      ? requestedGroups
      : selectedGroups(requestedGroups, `${where}.groups`);
  const exclude = exclusionDecisions(field(source, 'exclude'), where);
  const replace = replacementDecisions(field(source, 'replace'), where);
  for (const id of [...exclude.keys(), ...replace.keys()]) {
    const decision = exclude.has(id) ? 'exclude' : 'replace';
    ruleGroup(`${id}.md`, `${where}.${decision}.${id}`);
    if (exclude.has(id) && replace.has(id))
      return invalid(`${where}:${id}`, 'rule is both excluded and replaced');
  }
  return { name, repository, ref, parsedRef, groups, exclude, replace };
}

/**
 * Return validated configuration with sources, groups, and exception keys sorted.
 * Throws an invalid-input ValidationError for unsupported fields, malformed values, or conflicting declarations.
 * Rejects abbreviated-commit-shaped refs unless explicitly qualified as refs/tags/<name>.
 * This offline syntax check does not resolve refs or establish their existence in Git.
 */
export function configuration(input: unknown): ProjectConfig {
  const config = object(input, 'configuration');
  knownFields(
    config,
    ['schemaVersion', 'sources', 'localGroups'],
    'configuration',
  );
  if (field(config, 'schemaVersion') !== 1)
    return invalid('schemaVersion', 'only version 1 is supported');
  const repositories = new Set<string>();
  const sources: Array<LibrarySource> = Object.entries(
    object(field(config, 'sources'), 'sources'),
  )
    .sort(([a], [b]) => compare(a, b))
    .map(([name, raw]) => sourceConfiguration(name, raw, repositories));
  const localGroups = selectedGroups(
    field(config, 'localGroups'),
    'localGroups',
  );
  for (const id of localGroups) {
    if (
      sources.some(
        (source) =>
          typeof source.groups !== 'string' && source.groups.includes(id),
      )
    )
      return invalid(
        id,
        'an imported group cannot also be declared in localGroups',
      );
  }
  return { sources, localGroups };
}

File: internal/rules/configuration.go
// Parse project source declarations without resolving Git refs or reading files.

package rules

import (
	"encoding/json"
	"maps"
	"regexp"
	"slices"
	"strings"
)

// Configuration contains sources in alias order; an empty list is a local-only project.
type Configuration struct {
	Sources []Source `json:"sources"`
}

// Source declares one remote library and its selection and exception policy.
// Exactly one of Ref and Version is populated. ParsedRef is nil for version selections.
type Source struct {
	Name       string                 `json:"name"`
	Repository string                 `json:"repository"`
	Ref        string                 `json:"ref,omitempty"`
	Version    string                 `json:"version,omitempty"`
	ParsedRef  *GitRef                `json:"parsedRef,omitempty"`
	Groups     GroupSelection         `json:"groups"`
	Exclude    map[string]string      `json:"exclude"`
	Replace    map[string]Replacement `json:"replace"`
}

// Replacement refers to a local file and preserves the authored reason.
type Replacement struct {
	File   string `json:"file"`
	Reason string `json:"reason"`
}

var sourceNamePattern = regexp.MustCompile(`^[a-z][a-z0-9-]*$`)

// ParseConfiguration validates schema version 1, rejecting unknown fields and
// contradictory source declarations. Errors return no partial configuration.
// Strings retain authored whitespace; group IDs and source aliases are sorted.
func ParseConfiguration(input json.RawMessage) (Configuration, error) {
	fields, err := jsonObject(input, "configuration")
	if err != nil {
		return Configuration{}, err
	}
	if _, ok := fields["localGroups"]; ok {
		return Configuration{}, invalid("localGroups", "remove localGroups; local groups are discovered from local/<group>/_group.json")
	}
	if err := knownJSONFields(fields, []string{"schemaVersion", "sources"}, "configuration"); err != nil {
		return Configuration{}, err
	}
	var schema float64
	if json.Unmarshal(fields["schemaVersion"], &schema) != nil || schema != 1 {
		return Configuration{}, invalid("schemaVersion", "only version 1 is supported")
	}
	sources, err := jsonObject(fields["sources"], "sources")
	if err != nil {
		return Configuration{}, err
	}
	result := Configuration{Sources: make([]Source, 0, len(sources))}
	repositories := make(map[string]bool)
	for _, name := range slices.Sorted(maps.Keys(sources)) {
		source, err := parseSource(name, sources[name], repositories)
		if err != nil {
			return Configuration{}, err
		}
		result.Sources = append(result.Sources, source)
	}
	return result, nil
}

// parseSource validates fields in dependency order before checking exception conflicts.
func parseSource(name string, input json.RawMessage, repositories map[string]bool) (Source, error) {
	where := "sources." + name
	if !sourceNamePattern.MatchString(name) || name == "local" {
		return Source{}, invalid(where, "invalid or reserved source name")
	}
	fields, err := jsonObject(input, where)
	if err != nil {
		return Source{}, err
	}
	if err := knownJSONFields(fields, []string{"repository", "ref", "version", "groups", "exclude", "replace"}, where); err != nil {
		return Source{}, err
	}
	repository, err := jsonText(fields["repository"], where+".repository")
	if err != nil {
		return Source{}, err
	}
	address, err := ParseRepository(fields["repository"], where+".repository")
	if err != nil {
		return Source{}, err
	}
	if repositories[address.Identity] {
		return Source{}, invalid(where, "repository "+repository+" is declared more than once")
	}
	repositories[address.Identity] = true
	_, hasRef := fields["ref"]
	_, hasVersion := fields["version"]
	if hasRef == hasVersion {
		return Source{}, invalid(where, "specify exactly one of ref or version")
	}
	result := Source{Name: name, Repository: repository}
	if hasVersion {
		text, err := jsonText(fields["version"], where+".version")
		if err != nil {
			return Source{}, err
		}
		constraint, err := ParseVersionConstraint(text, where+".version")
		if err != nil {
			return Source{}, err
		}
		result.Version = constraint.String()
	} else {
		text, err := jsonText(fields["ref"], where+".ref")
		if err != nil {
			return Source{}, err
		}
		ref, err := ParseGitRef(text, where+".ref")
		if err != nil {
			return Source{}, err
		}
		result.Ref = text
		result.ParsedRef = &ref
	}
	result.Groups, err = ParseGroupSelection(fields["groups"], where+".groups")
	if err != nil {
		return Source{}, err
	}
	excludes, err := jsonObject(fields["exclude"], where+".exclude")
	if err != nil {
		return Source{}, err
	}
	result.Exclude = make(map[string]string, len(excludes))
	for _, id := range slices.Sorted(maps.Keys(excludes)) {
		text, err := jsonText(excludes[id], where+".exclude."+id)
		if err != nil {
			return Source{}, err
		}
		result.Exclude[id] = text
	}
	replacements, err := jsonObject(fields["replace"], where+".replace")
	if err != nil {
		return Source{}, err
	}
	result.Replace = make(map[string]Replacement, len(replacements))
	for _, id := range slices.Sorted(maps.Keys(replacements)) {
		replacement, err := parseReplacement(replacements[id], where+".replace."+id)
		if err != nil {
			return Source{}, err
		}
		result.Replace[id] = replacement
	}
	for _, id := range append(slices.Sorted(maps.Keys(result.Exclude)), slices.Sorted(maps.Keys(result.Replace))...) {
		decision := "replace"
		if _, ok := result.Exclude[id]; ok {
			decision = "exclude"
		}
		if _, err := GroupFromPath(id+".md", where+"."+decision+"."+id); err != nil {
			return Source{}, err
		}
		_, excluded := result.Exclude[id]
		_, replaced := result.Replace[id]
		if excluded && replaced {
			return Source{}, invalid(where+":"+id, "rule is both excluded and replaced")
		}
	}
	return result, nil
}

// parseReplacement checks path containment and local ownership without opening the file.
func parseReplacement(input json.RawMessage, location string) (Replacement, error) {
	fields, err := jsonObject(input, location)
	if err != nil {
		return Replacement{}, err
	}
	if err := knownJSONFields(fields, []string{"file", "reason"}, location); err != nil {
		return Replacement{}, err
	}
	file, err := jsonPath(fields["file"], location+".file")
	if err != nil {
		return Replacement{}, err
	}
	if !strings.HasPrefix(file, "local/") {
		return Replacement{}, invalid(location+".file", "replacement files must be under local/")
	}
	reason, err := jsonText(fields["reason"], location+".reason")
	if err != nil {
		return Replacement{}, err
	}
	return Replacement{File: file, Reason: reason}, nil
}
